packages/superset-ui-query/src/buildGroupedControls.ts

```diff
@@ -0,0 +1,38 @@
+import { ControlGroups, ResidualFormData } from './types/QueryFormData';
+import { GroupedControlData, ResidualGroupedControlData } from './types/ControlData';
+
+export default function buildGroupedControlData(
+  residualFormData: ResidualFormData,
+  controlGroups?: ControlGroups,
+): GroupedControlData & ResidualGroupedControlData {
+  const defaultedControlGroups: ControlGroups = {
+    /** These are predefined for backward compatibility */
+    metric: 'metrics',
+    percent_metrics: 'metrics',
+    metric_2: 'metrics',
+    secondary_metric: 'metrics',
+    x: 'metrics',
+    y: 'metrics',
+    size: 'metrics',
+    ...controlGroups,
+  };
+  const groupedControls: GroupedControlData & ResidualGroupedControlData = {
+    columns: [],
+    groupby: [],
+    metrics: [],
+  };
+
+  Object.entries(residualFormData).forEach(entry => {
+    const [key, residualValue] = entry;
+    if (Object.prototype.hasOwnProperty.call(defaultedControlGroups, key)) {
+      const controlGroup: string = defaultedControlGroups[key];
+      const controlValue = residualFormData[key];
```

Comment:
I'm a little confused by what this function is doing. Aren't `controlValue` and `residualValue` the same thing?

Reply:
Here's what I _think_ this function is essentially doing, correct me if I'm wrong?

```
const normalizedKey = Object.prototype.hasOwnProperty.call(defaultedControlGroups, key)
      ? defaultedControlGroups[key]
      : key;
 groupedControls[normalizedKey] = (groupedControls[normalizedKey] || []).concat(residualValue);
```

Reply:
I, too, was a bit puzzled here, so I've tried to write a simplified version of this block as well. Curious if this is missing something important (I'm a little tired, so... probably):
```
Object.keys(residualFormData).forEach(key => {
  const conformedKey = defaultedControlGroups[key] || key;
  groupedControls[conformedKey] = [...groupedControls[conformedKey] || [], residualFormData[key]]
});
```

Reply:
You're both right, this was poorly written 🤦 Will address.